streams/jvm/src/main/scala/zio/stream/experimental/platform.scala

```diff
@@ -119,4 +125,288 @@ trait ZStreamPlatformSpecificConstructors {
+    ZStream.fromJavaIteratorTotal(stream.iterator())
+
+  /**
+   * Create a stream of accepted connection from server socket
+   * Emit socket `Connection` from which you can read / write and ensure it is closed after it is used
+   */
+  def fromSocketServer(
+    port: Int,
+    host: Option[String] = None
+  ): ZStream[Any, Throwable, Connection] =
+    for {
+      server <- ZStream.managed(ZManaged.fromAutoCloseable(ZIO.effectBlocking {
+                  AsynchronousServerSocketChannel
+                    .open()
+                    .bind(
+                      host.fold(new InetSocketAddress(port))(new InetSocketAddress(_, port))
+                    )
+                }))
+
+      registerConnection <- ZStream.managed(ZManaged.scope)
+
+      conn <- ZStream.repeatEffect {
+                ZIO
+                  .effectAsync[Any, Throwable, UManaged[Connection]] { callback =>
+                    server.accept(
+                      null,
+                      new CompletionHandler[AsynchronousSocketChannel, Void]() {
+                        self =>
+                        override def completed(socket: AsynchronousSocketChannel, attachment: Void): Unit =
+                          callback(ZIO.succeed(Connection.make(socket)))
+
+                        override def failed(exc: Throwable, attachment: Void): Unit = callback(ZIO.fail(exc))
+                      }
+                    )
+                  }
+                  .flatMap(managedConn => registerConnection(managedConn).map(_._2))
+              }
+    } yield conn
+
+  /**
+   * Accepted connection made to a specific channel `AsynchronousServerSocketChannel`
+   */
+  class Connection(socket: AsynchronousSocketChannel) {
+
+    /**
+     * Read the entire `AsynchronousSocketChannel` by emitting a `Chunk[Byte]`
+     * The caller of this function is NOT responsible for closing the `AsynchronousSocketChannel`.
+     */
+    def read: ZStream[Any, Throwable, Byte] =
+      ZStream.unfoldChunkM(0) {
+        case -1 => ZIO.succeed(Option.empty)
+        case _ =>
+          val buff = ByteBuffer.allocate(ZStream.DefaultChunkSize)
+
+          ZIO.effectAsync[Any, Throwable, Option[(Chunk[Byte], Int)]] { callback =>
+            socket.read(
+              buff,
+              null,
+              new CompletionHandler[Integer, Void] {
+                override def completed(bytesRead: Integer, attachment: Void): Unit = {
+                  (buff: Buffer).flip()
+                  callback(ZIO.succeed(Option(Chunk.fromByteBuffer(buff) -> bytesRead.toInt)))
+                }
+
+                override def failed(error: Throwable, attachment: Void): Unit = callback(ZIO.fail(error))
+              }
+            )
+          }
+      }
+
+    /**
+     * Write the entire Chuck[Byte] to the socket channel.
+     * The caller of this function is NOT responsible for closing the `AsynchronousSocketChannel`.
```

Comment:
This again is a bit misleading for me, the caller already acquired the `Connection`, hopefully via using the `ZManaged` in the companion object, and will close it when the resource is released. This comment suggests for me that `write` itself closes the channel.